src/test/playwright/service-worker.spec.ts

```diff
@@ -0,0 +1,286 @@
+      <script type="sample/html" filename="index.html">
+        index.html
+      </script>
+    </playground-project>
+
+    <playground-preview project="p">
+    </playground-preview>
+  </body>
+</html>
+`;
+
+test.describe('service worker', () => {
+  let server: DevServer;
+  let indexUrl: string;
+  let originalSwCode: string;
+  let originalSwVersion: string;
+
+  /**
+   * When serving the service worker script, substitute the entire script for
+   * the given code.
+   */
+  const setServiceWorkerCode = (code: string) => {
+    swCode = code;
+  };
+  let swCode: string;
+
+  /**
+   * When serving the service worker script, substitute the given version in
+   * place of the real one.
+   */
+  const setServiceWorkerVersion = (version: string) => {
+    swVersion = version;
+  };
+  let swVersion: string;
+
+  /**
+   * When serving component code, substitute the given expected service worker
+   * version in place of the real one.
+   */
+  const setExpectedVersion = (version: string) => {
+    expectedVersion = version;
+  };
+  let expectedVersion: string;
+
+  const setServiceWorkerAndExpectedVersion = (version: string) => {
+    setServiceWorkerVersion(version);
+    setExpectedVersion(version);
+  };
+
+  /**
+   * Return a proimse that resolves the next time the service worker is served.
+   */
+  const serviceWorkerServed = () => serviceWorkerServedDeferred.promise;
+  let serviceWorkerServedDeferred = new Deferred<void>();
+
+  test.beforeAll(async () => {
+    originalSwCode = await readFile('./playground-service-worker.js', 'utf8');
+
+    const versionPattern =
+      /version:\s*(?<quote>["'`])(?<hex>[a-z0-9]+)\k<quote>/g;
+    const versionMatches = [...originalSwCode.matchAll(versionPattern)];
+    if (versionMatches.length !== 1) {
+      throw new Error(
+        'Expected 1 version string in playground-service-worker.js. ' +
+          `Found ${versionMatches.length}.`
+      );
+    }
+    const hex = versionMatches[0].groups?.hex;
+    if (!hex) {
+      throw new Error('Could not find "hex" capture group in regexp match.');
+    }
+    originalSwVersion = hex;
+
+    const plugin: Plugin = {
+      name: 'test-plugin',
+
+      serve(ctx) {
+        if (ctx.path === '/') {
+          return indexHtml;
+        }
+        return undefined;
+      },
+
+      transform(ctx) {
+        let body = ctx.body as string;
+        if (typeof body !== 'string') {
+          throw new Error('Expected string body');
+        }
+        if (ctx.path.endsWith('playground-service-worker.js')) {
+          body = swCode
+            .replace(originalSwVersion, swVersion)
+            // Extreme hack! We need some way to check which service worker
+            // served a response. Prepend the version into all responses.
+            .replace(/new Response\(/g, `new Response("${swVersion} "+`);
+          serviceWorkerServedDeferred.resolve();
+          serviceWorkerServedDeferred = new Deferred();
```

Comment:
This is re-assigned at the start of each test as well. Are both needed?

Reply:
It allows waiting for the service worker to be served multiple times within the same test. Granted we don't actually do that currently.